library/entity/interaction/src/main/java/org/quiltmc/qsl/entity/interaction/api/LivingEntityAttackEvents.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.entity.interaction.api;
+
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.damage.DamageSource;
+import net.minecraft.item.ItemStack;
+import org.quiltmc.qsl.base.api.event.Event;
+
+/**
+ * Contains events that invoke when a {@link LivingEntity} attacks another.
+ */
+public class LivingEntityAttackEvents {
+
+	/**
+	 * A callback that is invoked when a {@link LivingEntity} attacks another
+	 *     <strong>before</strong> the damage is dealt.
+	 * <p>
+	 * This event is cancellable by calling {@link DamageContext#cancel()}.
+	 * <p>
+	 * Implementations should not assume the hit will go through. Ideally, this
+	 *     event is used for conditionally cancelling or altering the damage value.
+	 */
+	public static final Event<Before> BEFORE = Event.create(Before.class,
+			callbacks -> context -> {
+				for (var callback : callbacks) {
+					callback.beforeDamage(context);
+
+					if (context.isCanceled()) return;
+				}
+			});
+
+	/**
+	 * A callback that is invoked when a {@link LivingEntity} attacks another
+	 *     <strong>after</strong> the damage is dealt.
+	 */
+	public static final Event<After> AFTER = Event.create(After.class,
+			callbacks -> (attacker, stack, target, source, damage) -> {
+				for (var callback : callbacks) {
+					callback.afterDamage(attacker, stack, target, source, damage);
+				}
+			});
+
+
+
+	@FunctionalInterface
+	public interface Before {
+		/**
+		 * Invoked <strong>before</strong> a {@link LivingEntity} damages another.
+		 *
+		 * @param context the {@link DamageContext} containing all the relevant arguments
+		 */
+		void beforeDamage(DamageContext context);
+	}
+
+	@FunctionalInterface
+	public interface After {
+		/**
+		 * Invoked <strong>after</strong> a {@link LivingEntity} damages another.
+		 *
+		 * @param attacker the attacking entity
+		 * @param stack the {@link ItemStack} in the attacker's main-hand
+		 * @param target the target {@link LivingEntity}
+		 * @param source the {@link DamageSource} of the attack
+		 * @param damage the damage that was dealt
+		 */
+		void afterDamage(LivingEntity attacker, ItemStack stack, LivingEntity target, DamageSource source, float damage);
```

Comment:
I wonder if there should be two `DamageContext`s. A `MutableDamageContext` for the before event, and an `ImmutableDamageContext` (names very much up for debate) for the after event. Then these interfaces can even be merged into one.

Reply:
So two more classes extending a abstract `DamageContext`? I think that can work.

Reply:
The separation of the `Before` and `After` functional interfaces I believe are more for clarification purposes however. The documentation could become a bit confusing if they are merged.

Reply:
I believe this would create a lot of duplicate code on the user's part. I think logic that determines cancellation or modifying damage would align a lot of the time and I'd rather not force users to duplicate code to perform these tasks.